### ISSUE TITLE
blocklist + fail2ban filter not blocking other path requests

### ISSUE THREAD

**Issue by rosskevin** [kickstarter/rack-attack#243]
I'm trying to setup/test for a set of penetration paths I've grepped from our logs.  I'm successful in blocking the ip for the given blocked path, but if you visit any other good path it is not blocked.

My goal is to block everything for that IP the moment a known penetration tester hits the API.

## Config

```ruby
  # --------------------------------------------------------------
  # HARD Block penetration testing paths or query strings.
  #
  # After *any* blocked requests in 5 minutes, block all requests from that IP for 60 minutes.
  Rack::Attack.blocklist('ip/penetration') do |req|
    # `filter` returns truthy value if request fails, or if it's from a previously banned IP so the request is blocked
    Rack::Attack::Fail2Ban.filter("penetration-#{req.ip}", maxretry: 0, findtime: 5.minutes, bantime: 60.minutes) do
      p = req.path
      # The count for the IP is incremented if the return value is truthy
      CGI.unescape(req.query_string) =~ %r{/etc/passwd} ||
        p.include?('/etc/passwd') ||
        p.match(/wp-admin/) ||
        p.match(/wp-login/) ||
        p.match(/\/pma/i) ||
        p.match(/\/db/) ||
        p.match(/\/sql/) ||
        p.match(/\/php/) ||
        p.match(/\/mysql/) ||
        p.match(/\/admin/) ||
        p.match(/\/database/) ||
        p.match(/\/2php/) ||
        p.match(/\/myadmin/i) ||
        p.match(/\/program/) ||
        p.match(/\/shopdb/)
    end
  end
```

## Rspec

```ruby
  it 'block by ip after first hit on penetration testing urls' do
    PENETRATION_PATHS.each do |path|
      Rails.cache.clear # clear last run

      assert_ok_favicon(CLIENT) # sanity

      # you banned sucka
      get path, {}, 'REMOTE_ADDR' => CLIENT
      assert_blocked("[#{path} 1]") # all good
      assert_blocked_favicon(1) # <=============== failure.  

      # travel past the ban time
      travel_to 61.minutes.from_now

      assert_ok_favicon(CLIENT)

      # you banned AGAIN sucka
      get path, {}, 'REMOTE_ADDR' => CLIENT
      assert_blocked("[#{path} 3]")
      get '/favicon', {}, 'REMOTE_ADDR' => CLIENT
      assert_blocked("[#{path} 4]")
      assert_blocked_favicon(1)

      travel_back
    end
  end

  def assert_blocked_favicon(i)
    get '/favicon.ico', {}, 'REMOTE_ADDR' => CLIENT
    assert_blocked("[/favicon.ico #{i}]")
  end

  def assert_blocked(i)
    expect(last_response.status).to eq(503), "Iteration #{i} should have been blocked but received #{last_response.status}"
  end
```

## TL;DR

Blocking is only working for the combination of IP and path.  I expect from this ruleset that once blocked, all requests will return a `blocked` response until the `bantime` expires.

Note that the above spec is fully working as long as I don't put in the `assert_blocked_favicon`.

Am I misunderstanding config or is this a bug?

**Comment by Novik** [comment]
Do you have an enabled cache?

**Comment by lmansur** [comment]
@rosskevin I can't reproduce this bug.

Are you sure you are using a valid cache store? By default Rails uses `ActiveSupport::Cache::NullStore` in test and development environments, which means nothing gets cached.

Try setting the cache store to something like `Rack::Attack.cache.store = ActiveSupport::Cache::MemoryStore.new` in your initializer file.

Let me know what you find.

**Comment by grzuy** [comment]
Closing for now.
We can reopen if we get more information from the user.

Thank you @Novik and @lmansur for helping out.

**Comment by rosskevin** [comment]
Sorry for the late response.  I did have cache enabled, but I confirmed this to be a user error in testing.  While I was setting:

```rb
Rack::Attack.cache.store = ActiveSupport::Cache::MemoryStore.new
```

in my test I was incorrectly clearing the wrong cache:

```rb
Rails.cache.clear
```

Fixed by just resetting the rack attack cache instead of looking to the global cache during specific rule testing.

Thanks to all for looking at it.

**Comment by grzuy** [comment]
@rosskevin Thank you very much for your update!
